### ISSUE TITLE
Android support for armv7/aarch32

### ISSUE THREAD

**Issue by ata-star** [simplex-chat/simplex-chat#538]
I have an LG Android device with Android 11

unable to use the app

**Comment by epoberezkin** [comment]
Thanks for reporting!

Could you please share some screenshot/errors you are seeing?

Do you by any chance have logcat viewer installed? If so, it would really help if you could share what is in the logs.

**Comment by ata-star** [comment]
![Screenshot_20220417-100918](https://user-images.githubusercontent.com/62115669/163706218-3e9fba25-86f8-4ec8-966c-65809ff4504d.png)

it says in german that the app is incompatible

same on Fdroid.

![Screenshot_20220417-101040](https://user-images.githubusercontent.com/62115669/163706258-ea31bc76-7c61-400f-b784-235f09b47781.png)

**Comment by ata-star** [comment]
could it be that you only compile for armv8

**Comment by epoberezkin** [comment]
Correct, the app does not support arm v7 CPUs, sorry about it.

**Comment by ata-star** [comment]
is it possible to compile it for v7 androids or is it a deeper problem?

**Comment by epoberezkin** [comment]
I don't think we can - Haskell has aarch64 support, but not the previous ARM versions...

**Comment by ata-star** [comment]
thats funny

**Comment by epoberezkin** [comment]
Indeed, it's a problem, but we hope to at least support Android v8/9 - that would almost 1.5x the number of supported devices.

In any case, both these problems will be solved by time...

**Comment by ata-star** [comment]
my phone is not even old, its 1 year old... has android 11

**Comment by abrasmag3** [comment]
Rewriting the codebase in Rust is such a time consuming task that it is really frustrating to recommend it to improve compatibility.

However, there are some nice things about rust that i will note.
- Protocols usually perform best in a low level environment (citation needed)
- Rust excels in concurrency and has STM implementations based closely on Haskell's implementations
- Rust is often very low latency and has a small to medium file size

I don't really think this is enough to focus all effort on rewriting the project in rust, but I think a rust implementation of this protocol would be really cool to see. I may try to do so eventually.

Also, this is off topic for the reason stated in the first sentence.

**Comment by ata-star** [comment]
> I don't think we can - Haskell has aarch64 support, but not the previous ARM versions...

Are you sure?
i think it does...

**Comment by ata-star** [comment]
Bump

 haskell has armv7 support, port the app please?

**Comment by epoberezkin** [comment]
Right, this is indeed probably nix and Haskell problem…

**Comment by ltGuillaume** [comment]
Is this a short-term goal? Seems you can get there with QEMU pretty easily.

Not supporting ARMv7 rules out a whole batch of devices, so it'll take a lot more time for SimpleX to gain traction.

**Comment by ata-star** [comment]
Or even gain traction at all

**Comment by ata-star** [comment]
i see no future in this project if i cant even run it on a modern phone where every other app works
(Element/Matrix or XMPP)

Why should anyone use this if you cant be sure it works on all modern devices

**Comment by epoberezkin** [comment]
1. It’s definitely a priority to run on aarch32 / ARM v7

**Comment by epoberezkin** [comment]
> Is this a short-term goal? Seems you can get there with QEMU pretty easily.

that’s not that easy, but we are doing it.

**Comment by why-not-try-calmer** [comment]
My phone belongs to the armeabi v7a platform. I can run any Googe Play app and use any instant messaging application. But cannot use SimpleX,to my great dismay.

**Comment by ltGuillaume** [comment]
> 1. It’s definitely a priority to run on aarch32 / ARM v7, but a lower priority than to run on Android 8 and 9 which is likely to happen sooner.

I really don't get this. Why would it be more important to be able to run the app on an end-of-life and wholly unsupported operating system than to be able to run it on up-to-date devices (considering the software) that happen to have a different architecture?

**Comment by epoberezkin** [comment]
> Why would it be more important to be able to run the app on end-of-life and wholly unsupported operating system versions

I might be wrong, but it seems that it gives more value (= more supported devices) for less effort... According to Google ~20-25% of devices are still running Android 8 and 9... I don't know the exact number, but it seems that much less than 20% of devices have 32-bit only CPUs, and at the same time it is substantially more complex to support – that's why it is a lower priority. 

Also, support for 32-bit CPU seems dying faster than Androids 8 and 9, and some 64 bit devices stopped supporting 32-bit apps (again I might be wrong here) - https://www.androidauthority.com/arm-32-vs-64-bit-explained-1232065/.

If you have a better statistics we might be reconsidering the priorities, but it's in progress anyway – 32bit support is next on the list after supporting Android 8 and 9.

**Comment by ltGuillaume** [comment]
> According to Google ~20-25% of devices are still running Android 8 and 9...

There will undoubtedly be a significant amount of overlap between devices on Android 8/9 and 32-bit architectures. So fixing support for Android 8/9 will only have true meaning if you'll also support 32-bit CPU's.

> Also, support for 32-bit CPU seems dying **faster than Androids 8 and 9**, and some 64 bit devices stopped supporting 32-bit apps (again I might be wrong here) - https://www.androidauthority.com/arm-32-vs-64-bit-explained-1232065/.

It can't, because there is no support for Android 8 and 9 anymore (https://endoflife.date/android).

**Comment by ata-star** [comment]
I HAVE AN ANDROID 11 WITH ARMv7 

God Damn it the version is unimportant  JESUS!

**Comment by ata-star** [comment]
@ltGuillaume 
There are MANY new Phones produced like mine that has Android 11 AND armv7

**Comment by why-not-try-calmer** [comment]
> A credible statistics on the distribution would be very helpful. Play Store must have it, but I can't find it...
There you go: https://gs.statcounter.com/os-version-market-share/android
We can discuss all day long how "credible" this is (consolidated ~22% market share between 9 & 8) but I hope there's common sense in the idea that since Play Store supports ALL devices running Android 9, they must do it for a reason that should also apply to SimpleX (relative to a reasonable cost in terms of time and effort).

**Comment by epoberezkin** [comment]
@why-not-try-calmer this one I've seen, but I meant the stats specifically on the share of 32 vs 64 bits devices, and maybe with versions as the second dimension. Having visibility of it sooner would have helped, it does seem we were underestimating the share of 32bit devices.

> consolidated ~22% market share between 9 & 8

this seems consistent with google stats.

> Also let's be real: looking from the code there's 0 reason to think that Simplex is too demanding CPU-wise or GPU-wise or RAM-wise;

Huh, I wish it was the case 100% of the time :) We've seen cases when because of uncapped retries it was becoming a traffic hog, but we are discovering such issues quite rarely now... Memory wise it's quite conservative - I have about 400 contacts and I've not seen it going over 200-300Mb, only when you scroll very far into the chat history it goes higher (because it's only one-way infinite scrolling).

> so it's always a good idea to cater for sober users who like to stick to simple, affordable devices -- if possible.

That I agree with. I know people succeeded running a server on armv7, even though having it on Android adds some extra variables to compilation process, we might succeed with it, fingers crossed.

**Comment by dayfuaim** [comment]
And what about Android 7 (aarch64)?

**Comment by ltGuillaume** [comment]
Any news on this? Just tried it on someone else's device, dying to test it properly on my own mobile.

**Comment by jnolan7599** [comment]
@epoberezkin 

So this 'issue' has been known for now nearly 10 months? It seems to rather speak to the priorities of this endeavour of yours, doesn't it?

In 2020 the [Alcatel 1B](https://www.gsmarena.com/alcatel_1b_(2020)-10019.php) was released with its quad core Cortex-A53.  A Android Go 10 device that uses _armeabi-v7a_... so a 64-bit SOC running a 32-bit OS. I think it behoves SimpleX to keep in mind not everyone cares about chasing the latest consumer trend(s) of throwing good money after money on the latest shiny device being dangled to be pumped, dumped out on the market like slop in a trough.

Ref: #1726

**Comment by epoberezkin** [comment]
> So this 'issue' has been known for now nearly 10 months? It seems to rather speak to the priorities of this endeavour of yours, doesn't it?

It not about just priorities, it's a complex issue to resolve, as it is a GHC compiler limitation - it only relatively recently started supporting aarch64 CPUs and has bugs on aarch32... We underestimated how widespread 32bit CPUs are, as you have seen lots of 64bit CPUs are locked into 32bit mode and it not available in any stats - I still don't know what share of devices we are talking about. Although I am not sure what we would have done if we estimated it correctly...

**Comment by jnolan7599** [comment]
@epoberezkin 

I can somewhat sympathize but from an outsider's perspective it leaves an incredibly bad taste in the mouth. I'm now unable to introduce SimpleX to those know they should be moving to privacy/encrypted applications but lack the technical skill to even sideload an apk... which is about where are my Android skills for software development.  They already have phones under carrier contract(s). They're _certainly_ not going to buy a flagship device for what is of little distinction, to them, of a replacement to Facebook/Meta Messenger.

Has your group considered reaching out to other shops/start-ups who use that compiler & perhaps pool some resources to get those 'bughunted'? That's merely a suggestion from some so frustrated by such a promising app.

**Comment by epoberezkin** [comment]
I added some comment there. I understand the frustration.

You don't need a flagship device, there is a large number of low end 64bit devices – e.g., Samsung A3.

> Has your group considered reaching out to other shops/start-ups who use that compiler & perhaps pool some resources to get those 'bughunted'? That's merely a suggestion from some so frustrated by such a promising app.

We happen to be, probably, the most advanced Haskell mobile app – it's not that common to build mobile apps in Haskell (I don't know any other anywhere near as complex app built in Haskell). And we do have some good help at the moment...

**Comment by jnolan7599** [comment]
@epoberezkin FYI: I think I was editing mine while you posted your latest.

**Comment by epoberezkin** [comment]
yep :) no problem.

**Comment by jnolan7599** [comment]
@epoberezkin ... then I'll just have to sit & stew... but before that do us all a favor & get an RSS feed up for the Simplex blog, FFS. Not everyone knows how to convert GitHub URLs to RSS/atom feeds. The blog articles are far more interesting than a mere changelog.

**Comment by epoberezkin** [comment]
Yep, sorry. We do post all updates/releases to reddit and mastodon - once we have 32bit support that'll be a big news, trust me :) Also you can connect to the users group via CLI in the meanwhile and vent there too - there will be lots of sympathy I am sure ;) It is very annoying...

**Comment by jnolan7599** [comment]
@epoberezkin Your good nature is appreciated. This is the most I care to vent ATM. I'll add another feed for SimpleX's reddit though I refuse to subscribe given Reddit's well known history of censorship.

I know you're already aware of the impact your app can have on high stakes whistleblowing & a journalist's source confidentiality in the face of well funded bad actors.

**Comment by epoberezkin** [comment]
> I know you're already aware of the impact your app can have on high stakes whistleblowing & a journalist's source confidentiality in the face of well funded bad actors.

That's good to hear. I have a huge respect to a journalistic profession, and if we can make their job protecting sources a bit easier (you can't disclose what you don't know) - that is great.

**Comment by jnolan7599** [comment]
@epoberezkin 

Great to hear.

When time permits I intend to add a fully _non-logging_ relay to the STUN/TURN pool for a similar rational. On that note it's rather a shame Syncthing's [relay network](https://docs.syncthing.net/users/strelaysrv.html) couldn't somehow be utilized for that function or as complimentary to it. They already have infrastructure in place for NAT transversal.. [420 instances](https://relays.syncthing.net/) could add some nice 'noise' to the signal.

That's all well off topic though so I'll stop wasting your time. Regards.

**Comment by why-not-try-calmer** [comment]
That's the point where I unsubscribe from this topic.

**Comment by ltGuillaume** [comment]
> https://www.reddit.com/r/SimpleXChat/comments/10dcyze/simplex_chat_lack_of_android_32_bit_cpu_support/ - some interesting comments to this post.
> 
> TL;DR: the advice is to upgrade to aarch64 devices for security reasons. It doesn't mean we won't support 32bit, as most people won't follow this advice, but somebody might I think...

This may sound harsh, but I'd rather see your being honest.

It's very, _very_ strange to talk about "security reasons" and at the same time about supporting Android 8 and 9 (it even being a higher priority). You may as well just be honest and state that adding support for 32-bit, up-to-date, devices is more work than supporting earlier _EOL_ and with certainty non-secure Android versions, instead of hiding behind security reasons...

**Comment by x80486** [comment]
I don't think anyone is hiding behind anything. They stated clearly that they are working to support the architecture (makes no sense to be honest) and since it's open source, people should be able to watch the progress at some point — if not already. In the meantime there is nothing wrong providing evidence that it can have downsides around security and what not. This looks very much like a side conversation, not as something that's being put in front of the development process to delay it.

New projects require _some magic_ to succeed because time and money are always limited. I would rather see `SimpleX` spending time on the future, fixing existing bugs, and solidifying what's done already, rather than trying to support something that will be discarded any time soon.

**Comment by ltGuillaume** [comment]
That's fine, but if you say it's because of security reasons, then drop possible support for Android 9 and below, too.

That being said, offering money to potential users for making a switch is something I'd never have expected, lolz.

**Comment by epoberezkin** [comment]
> @ltGuillaume Android 8 and 9 (it even being a higher priority). You may as well just be honest and state that adding support for 32-bit, up-to-date

We actually did re-assess it as a result of this conversation, and the work to add 32bit support is on-going. That I made some comment several months ago doesn't mean I still think the same :)

We're not hiding behind security reasons, that was not the point of sharing this comment.

> @x80486 They stated clearly that they are working to support the architecture (makes no sense to be honest)

I thought that too. But the logic is very simple here - not supporting 32 bit architecture doesn't reduce the network size proportionally to the share of 32 bit devices, if it were the case it would make no sense indeed. It has a disproportionately higher impact on network size, because of its impact on network density. Imaging a group of 4 close friends with only one of them having 32 bit device... The result is that none of them will use SimpleX Chat, even if 3 of them want to use it. Anecdotally, I have such an active group myself, with one of key members having 32 bit device (and yes, I suggested I'll just buy the device, but we're still talking with this particular group elsewhere). My estimate is that adding support for 32bit devices will 2-3x the network size, because of disproportionate impact of network density on network size.

> That's fine, but if you say it's because of security reasons, then drop possible support for Android 9 and below, too.

We're not looking into Android 9 ATM, work on 32 bit support is on-going - there is a substantial progress.

> That being said, offering money to potential users for making a phone switch is something I'd never have expected, lolz.

We do have a small marketing budget, I'd rather give it to the users :)))

**Comment by ltGuillaume** [comment]
Thanks for the reply, great to see these developments (no Android 8/9 plans, but 32-bit progress), I think it's the right call, and your reasoning about the impact on network size (in the current "landscape") seems sound.

**Comment by epoberezkin** [comment]
@ltGuillaume @x80486 @why-not-try-calmer @ata-star what specific 32 bit device would you recommend to test the app on? Can't run 32 bit on Google pixel it seems...

**Comment by ata-star** [comment]
i have my LG K41 something, it is an 32bit device

**Comment by why-not-try-calmer** [comment]
> @ltGuillaume @x80486 @why-not-try-calmer @ata-star what specific 32 bit device would you recommend to test the app on? Can't run 32 bit app on Google pixel it seems...

I wish I could help with testing but I am using an Android 9 phone.

**Comment by ltGuillaume** [comment]
Do you have any old device lying around? Doesn't need an up-to-date stock OS, with e.g. https://divestos.org/pages/devices you'll be reasonably up-to-date (11/12/13) with additional security over LineageOS and even backported kernel security patches.

**Comment by tomquas** [comment]
i recently bought a samsung a13, guess what, i was surprised by the missing support.
this phone is running android 12, and is nominated for an update to android 13 within the coming weeks.

it is not supported by simplex chat.

for us, it's entirely about market share of the corresponding devices – and that make armeabi v7 platform a MUST have. 19% market share represents several tens of millions devices.

source: https://android.stackexchange.com/questions/248820/how-is-the-market-divided-between-android-cpu-architectures
 
@epoberezkin can you say something about ETA? this thread is older than 10 months now
thx

**Comment by tomquas** [comment]
@epoberezkin rewrite in rust ;)

seriously: google can warn about 32-bit apps as much as they want, 32-bit devices will stay in use for longer than we can expect.

do you have an ETA? market share seems still ~20%, representing tens of millions of devices...

**Comment by epoberezkin** [comment]
We have a working build for ARM v7a working on Android 8+ - I will upload here to test in a day or two.

**Comment by epoberezkin** [comment]
Happy to see that many upvotes on this news in less than a day :)

The experimental APK for 32bit devices with ARM v7a CPU is here: https://github.com/simplex-chat/simplex-chat/releases/tag/v4.5.3

It should support Android 8+/API 26+ (the currently released version supports Android 10+/29+)

We did fix several limitations of lower android versions we could find, and we could not find any CPU-specific issues. But given that it's literally world's first Haskell build for ARM v7a CPU, it's reasonable to expect bugs, so definitely don't use it with your existing main chat profile yet.

**Comment by tomquas** [comment]
thank you so much! am making an exception to the rule and am jumping in to test the world's first...

**Comment by epoberezkin** [comment]
@tomquas which rule? :)

**Comment by tomquas** [comment]
first feedback: 

on a galaxy a13 (SM A135F) running android 12 i get the following error...
```
E/flutter (14622): [ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: Bad state: Invalid argument(s): Failed to load dynamic library 'libsimplex.so': dlopen failed: cannot locate symbol "__svfscanf" referenced by "/data/app/~~lXL6-_sZj2UUHFapIDdwRw==/com.wildtrolls.u2nite-ztdnoT7Is2ekAwGJFgBRNg==/base.apk!/lib/armeabi-v7a/libsimplex.so"...
```

both required libs reside under
```
android/app/src/main/jniLibs/armeabi-v7a:
.rw-r--r-- tom staff 116 MB Wed Mar  8 13:54:57 2023 libsimplex.so
.rw-r--r-- tom staff  12 MB Wed Mar  8 13:54:57 2023 libsupport.so
```

any hints appreciated...

**Comment by epoberezkin** [comment]
I don't understand... What file did you install?

**Comment by avently** [comment]
@tomquas I have devices on Android 13 and 9. Both works fine with provided apk. 

Do you use the same apk or your specific build of it?

Logs you attached have a note about flutter. We don't use flutter in the app. Where it comes from? Did you use the libs in your own build based on flutter or is it come from system that uses flutter for some reason?

**Comment by tomquas** [comment]
sorry for commenting a bit out of context: we're developing a flutter app and integrate simplex chat using libsimplex.so, libsupport.so, and some dart/flutter glue code.

so i didn't install the apk on the phone but extracted both libs and threw it into our own project. during runtime, while dart trying to load libsimplex.so, we see the failure log shown above. does this clarify things?

**Comment by epoberezkin** [comment]
it won't work without changes in the app itself

**Comment by avently** [comment]
@tomquas currently, the libs doesn't have some symbols required for linking. So to make it working, I added these symbols to a linking stage, see here for more info:
https://github.com/simplex-chat/simplex-chat/commit/77d7c8ab93436dbf1fb30a1f824869902470405d

https://github.com/simplex-chat/simplex-chat/commit/af44e03d4ea582f8baef7e7c181ea5b8c133b7b8

Maybe later we will include required symbols as a stubs to the libs code, so the process in commits will not be needed

**Comment by tomquas** [comment]
> Maybe later we will include required symbols as a stubs to the libs code, so the process in commits will not be needed

awesome, thanks a lot. i'll be watching this thread and will be happy to report more results...

**Comment by epoberezkin** [comment]
https://github.com/simplex-chat/simplex-chat/releases/tag/v4.6.0-beta.0

This beta supports armv7a and Android 8+

**Comment by Kortik7** [comment]
> 

Installed 4.6.0-beta.0 on pixle 3a with GrapheneOS same symptoms as before. app at opening just closes and doesn't go anywhere. closes back to main page.

**Comment by epoberezkin** [comment]
If it persists with 4.6 please open another issue. This one is done!
